### PR TITLE
Migrating part of ZIOSpecJvm to zio-test.

### PR DESCRIPTION
**Pull request by LGLO**
Migrated random part of ZIOSpecJvm to ZIOSpec, using zio-test

### CODE REVIEW

File: core-tests/shared/src/test/scala/zio/ZIOSpec.scala [hunk 1]
```diff
@@ -68,6 +109,44 @@ object ZIOSpec
...
+          },
+          testM("on failure") {
+            assertM(ZIO.fail("Fail").left.either, isLeft(isSome(equalTo("Fail"))))
+          }
+        ),
+        suite("leftOrFail")(
+          testM("on Left value") {
+            assertM(UIO(Left(42)).leftOrFail(ExampleError), equalTo(42))
+          },
+          testM("on Righ value") {
```
Comment (mijicd) [inline core-tests/shared/src/test/scala/zio/ZIOSpec.scala]: Minor typo here.

File: core-tests/shared/src/test/scala/zio/ZIOSpec.scala [hunk 2]
```diff
@@ -1268,3 +1388,48 @@ object ZIOSpecHelper {
 
   def AsyncUnit[E] = IO.effectAsync[E, Unit](_(IO.unit))
 }
+
+object UncurriedBracketCompilesRegardlessOrderOfEAndRTypes {
```
Comment (mijicd) [inline core-tests/shared/src/test/scala/zio/ZIOSpec.scala]: Can this be moved outside of spec file?

File: core-tests/shared/src/test/scala/zio/ZIOSpec.scala (hunk 1 above)
Comment (LGLO) [reply core-tests/shared/src/test/scala/zio/ZIOSpec.scala]: Will fix!

File: core-tests/shared/src/test/scala/zio/ZIOSpec.scala (hunk 2 above)
Comment (LGLO) [reply core-tests/shared/src/test/scala/zio/ZIOSpec.scala]: Totally seprate file? BracketTypeInferenceSpec.scala?

Comment (mijicd) [reply core-tests/shared/src/test/scala/zio/ZIOSpec.scala]: Yeah. Not that it matters much, but it's a bit awkward check to mix with regular specs.

**Comment by ghostdogpr** [comment]
Nice!

**Pull request merged**